[AIRFLOW-3632] Allow replace_microseconds in trigger_dag REST request

acroos opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3632

### Description

Reads the value for `replace_microseconds` from the API's `trigger_dag` endpoint.  This will allow the endpoint to trigger > 1 dag per second.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
- `TestApiExperimental.test_trigger_dag_for_date_with_replace_microseconds_false`

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

Commit:
[AIRFLOW-3632] Add replace_microseconds to trigger_dag endpoint

Enable the caller of the trigger_dag API endpoint to specify a
boolean value for replace_microseconds.  When false, this will
store microseconds in the database for the execution_id, thus allowing
more than 1 request per second to be processed by the REST API

ashb commented on airflow/www/api/experimental/endpoints.py:
```diff
@@ -74,8 +75,12 @@ def trigger_dag(dag_id):
+
     try:
-        dr = trigger.trigger_dag(dag_id, run_id, conf, execution_date)
+        dr = trigger.trigger_dag(dag_id, run_id, conf, execution_date, replace_microseconds)
```
This would fit the goal, but I feel there is a better way of expressing this.

If you follow all the code through I think what we should do is: if an explicit timestamp is passed in, we should use that exactly. If one is not passed then the current default of "generate the execution date from now(), but truncate it to nearest whole seconds" come in to play.

I think that would lead to a clearer to understand API.

WDYT?

acroos replied:
yeah that seems much nicer.  I felt like passing `'replace_microseconds': 'false'` in the request body felt a bit odd, but didn't want to make _too_ big of a change to start with.  I'll make the changes in the morning

acroos replied:
upon second thought... is there a particular reason that the time _needs_ to be truncated?  having a default behaviour so limiting seems like a bug

acroos replied:
I found [this commit](https://github.com/apache/airflow/commit/b7c828bf094d3aa1eae310979a82addf7e423bb0) from Feb 2017 that mentions why microseconds were replaced at that time.  I would like to think that isn't still relevant, but I'm not sure to be honest.

ashb replied:
It doesn't seen needed, other than it's sometimes "nice" to see whole seconds. For the API it is probably more sensible to default to un-trucated time, but this would be potentially "breaking" change, so we need to think about how we upgrade existing users.

For the 1.10 release series I would probably suggest that the default shouldn't change, just in case someone is relying on it.

If you could do this as two PRs (both against master) the first adding the behavour we talked about, but not changing the deafult behaviour (which we can get in to the next 1.10.x release, I will cherry-pick the merged PR back to the relase branch) and then once that is merged a second one that changes the default to be as you describe.

How does that sound?

acroos replied:
sounds great, will do!

Commit:
[AIRFLOW-3632] Only replace microseconds if execution_date is None

No need to require a user to pass in replace_microseconds to the
request body; instead we should use exactly the date that is given.
We will still replace the microseconds on execution_date if none is
passed in (and the param is True, which is the default)

ashb commented on airflow/api/common/experimental/trigger_dag.py:
```diff
@@ -107,7 +107,7 @@ def trigger_dag(
     :param conf: configuration
     :param execution_date: date of execution
-    :param replace_microseconds: whether microseconds should be zeroed
+    :param replace_microseconds: whether microseconds should be zeroed (ignored if execution_date is given)
```
I'm not sure if it matters, but the view in `airflow/www/api/experimental/endpoints.py` doesn't pass through a replace_microseconds parameter, so there's no way of setting this to anything that the default True from the REST API

acroos replied:
that's true, maybe would be nice to add that as well.

ashb reviewed (commented):
Code I think looks good, but we need a note in UPDATING.md about this change.

Is it possible to get the old behaviour back by explicitly passing replace_microsecond=True in the API call? If not could you please add a way to get the old behaviour back just in case someone is depending upon it.

Commit:
[AIRFLOW-3632] Re-allow replacing microseconds with execution_date

A user can now choose to pass the execution_date in but still have
the microseconds replaced by also sending replace_microseconds=true
in the request body

Commit:
[AIRFLOW-3632] Test _almost_ all scenarios with microsecond replacing

The only way that replace_microseconds should be accessible
is through the endpoints, so leave the testing there

ashb commented on UPDATING.md:
```diff
@@ -41,6 +41,14 @@ assists users migrating to a new version.
+The default behavior was to strip the microseconds (and milliseconds, etc) off of all dag runs triggered by
+by the experimental REST API.  The default behavior will change when an explicit execution_date is
+passed in the request body.  It will also now be possible to have the execution_date generated, but
+keep the microseconds by sending `replace_microseconds: false` in the request body.  The default
```
```suggestion
keep the microseconds by sending `replace_microseconds=false` in the request body.  The default
```

acroos replied:
👍 fixed

ashb commented on UPDATING.md:
```diff
@@ -41,6 +41,14 @@ assists users migrating to a new version.
+by the experimental REST API.  The default behavior will change when an explicit execution_date is
+passed in the request body.  It will also now be possible to have the execution_date generated, but
+keep the microseconds by sending `replace_microseconds: false` in the request body.  The default
+behavior can be overridden by sending `replace_microseconds: true` along with an explicit execution_date
```
```suggestion
behavior can be overridden by sending `replace_microseconds=true` along with an explicit execution_date
```

acroos replied:
👍 fixed

Commit:
[AIRFLOW-3632] Fix format of replace_microseconds in updating doc

ashb commented on UPDATING.md:
```diff
@@ -93,6 +93,28 @@ delete this option.
+keep the microseconds by sending `replace_microseconds=false` in the request body.  The default
+behavior can be overridden by sending `replace_microseconds=true` along with an explicit execution_date
+
+### Additional arguments passed to BaseOperator cause an exception
```
Looks like you've got some other changes included here?

acroos replied:
hmm apparently?  I guess when rebasing somehow picked up extra commits

acroos replied:
fixed
